### ISSUE TITLE
Add types/JSDoc

### ISSUE THREAD

**Issue by uncenter** [11ty/eleventy-img#206]
I think users having to install a separate package for typed usage could be improved upon by simply including types here. See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/11ty__eleventy-img/index.d.ts.

**Comment by SethFalco** [comment]
It wouldn't be a huge amount of work, but enough that I think I should ask before I do it.

I'd love to see the library bundle its types as well. Doing it through JSDocs is cool, and will surface up some of the existing "docs" (code comments) that you have that probably should've been in a JSDoc anyway. ^-^'

[Svelte](https://svelte.dev/docs/kit/packaging#TypeScript) is a great reference for adding types the right way to a library _(imo)_. SVGO could be used as a reference too (though Svelte's workflow is much more polished).

TL;DR for SVGO, since I maintain it:

- SVGO is a **_JavaScript_** project, not a TypeScript project. We define types through JSDocs, and try to keep TypeScript <abbr title="Domain-specific language">DSL</abbr> minimal.
- We validate our own code with tsc (which natively supports JavaScript + JSDoc)!
- We emit type declarations (`.d.ts`) from our JSDocs in a build step. Note, SVGO is still 100% usable without a build step, we aren't transpiling the project to JavaScript. The build step is just for tsc to generate `*.d.ts` files with types derived from our JSDocs, which we include in our npm package.

Fortunately, a lot of the work has already been done. Since `@types/11ty__eleventy-img` already exists, it can largely be incorporated into this project. (Preserving authorship/attribution with `--author` ofc!)

In my opinion, the best way to incorporate it would be to convert the contents of `@types/11ty__eleventy-img` to JSDocs and annotate your exports. For example:

**`src/image.js`**
```js
/**
 * @typedef {"webp" | "jpeg" | "png" | "svg" | "avif"} ImageFormat
 *
 * @typedef {object} ImageOptions
 * @property {Readonly<(number | "auto" | null)[]>} [widths]
 *   Controls how many output images will be created for each image format.
 *   Aspect ratio is preserved.
 * @property {ImageFormat[]} [formats] Controls the output image formats.
 * @property {string[]} [formatFiltering]
 * @property {number} [concurrency]
 * @property {string} [urlPath]
 *   A path-prefix-esque directory for the `<img src>` attribute. e.g. `/img/`
 *   for `<img src="/img/MY_IMAGE.jpeg">`.
 * @property {string} [outputDir]
 *   Where to write the new images to disk. Project-relative path to the
 *   output image directory. Maybe you want to write these to your output
 *   directory directly (e.g. `./_site/img/`)?
 * …
 */

export default class Image {
  // …

  /**
   * @param {string | URL | Buffer} src
   * @param {ImageOptions} [options]
   */
  constructor(src, options = {}) {
    // …
  }
  // …
}
```

Then the plumbing:

- adding `typescript` to `devDependencies`
- personally I'd recommend `tsd` or `vitest` which can test that your types behave and narrow in the way that you'd expect. Though just using the built-in `satisfies` is a more lenient but fine approach too if you don't want the extra dependencies.
- exporting the `types` field in `package.json`

This allows type checking downstream, and surfaces much more documentation to users. (i.e., during code completion or on hover)

Would the maintainers have any qualms with a PR that took this approach? Or did you perhaps already have a direction in mind? (Ofc feedback / iteration is welcome.)

Edit: Ahh… I just realized this issue was opened _by an org member_. I'll assume it's fine then!
